src/libsystemd-network/sd-dhcp-server.c

```diff
@@ -712,7 +715,54 @@ static int get_pool_offset(sd_dhcp_server *server, be32_t requested_ip) {
         return be32toh(requested_ip & ~server->netmask) - server->pool_offset;
 }
 
-static int dhcp_server_relay_message(sd_dhcp_server *server, DHCPMessage *message, size_t opt_length) {
+static size_t relay_option_length(sd_dhcp_server *server) {
+        size_t length = 0;
+        if (server->agent_circuit_id)
+                length += strlen(server->agent_circuit_id) + 2;
+        if (server->agent_remote_id)
+                length += strlen(server->agent_remote_id) + 2;
+
+        return length;
+}
+
+static uint8_t* append_suboption(uint8_t* dst, uint8_t subcode, char* value, uint8_t length) {
+        dst[0] = subcode;
+        dst[1] = length;
+        memcpy(dst + 2, value, length);
+        return dst + 2 + length;
+}
+
+static int append_agent_information_option(sd_dhcp_server *server, DHCPMessage *message, size_t msg_len, size_t buf_len) {
+        assert(message);
+
+        uint8_t* end_option_ptr = memrchr(message->options, SD_DHCP_OPTION_END, msg_len - sizeof(DHCPMessage));
+        if (end_option_ptr == NULL)
+                return -1;
+
+        size_t options_length = end_option_ptr - message->options;
+        uint8_t inf[(255 + 2)*2];
```

Comment:
Hm, the buffer size must be equal or smaller than `255`,

Reply:
Right, this is a bug.

Reply:
Fixed, and added enum for the supotion codes.